pkg/controller/direct/bigquerydataset/dataset_controller.go

```diff
@@ -0,0 +1,323 @@
+
+	// Convert KRM object to proto message
+	desiredKRM := a.desired.DeepCopy()
+	desired := BigQueryDatasetSpec_ToAPI(mapCtx, &desiredKRM.Spec, desiredKRM.Name)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	resource := clone.Clone(a.actual).(*api.Dataset)
+
+	// Check for immutable fields
+	if !reflect.DeepEqual(desired.Location, resource.Location) {
+		return fmt.Errorf("BigQueryDataset %s/%s location cannot be changed, actual: %s, desired: %s", u.GetNamespace(), u.GetName(), resource.Location, desired.Location)
+	}
+
+	// Find diff
+	updateMask := &fieldmaskpb.FieldMask{}
+	if !reflect.DeepEqual(desired.Description, resource.Description) {
+		resource.Description = desired.Description
+		updateMask.Paths = append(updateMask.Paths, "description")
+	}
+	if !reflect.DeepEqual(desired.FriendlyName, resource.FriendlyName) {
+		resource.FriendlyName = desired.FriendlyName
+		updateMask.Paths = append(updateMask.Paths, "friendly_name")
+	}
+	if !reflect.DeepEqual(desired.DefaultPartitionExpirationMs, resource.DefaultPartitionExpirationMs) {
+		resource.DefaultPartitionExpirationMs = desired.DefaultPartitionExpirationMs
+		updateMask.Paths = append(updateMask.Paths, "default_partition_expirationMs")
+	}
+	if !reflect.DeepEqual(desired.DefaultTableExpirationMs, resource.DefaultTableExpirationMs) {
+		resource.DefaultTableExpirationMs = desired.DefaultTableExpirationMs
+		updateMask.Paths = append(updateMask.Paths, "default_table_expirationMs")
+	}
+	if !reflect.DeepEqual(desired.DefaultCollation, resource.DefaultCollation) {
+		resource.DefaultCollation = desired.DefaultCollation
+		updateMask.Paths = append(updateMask.Paths, "default_collation")
+	}
+	if desired.DefaultEncryptionConfiguration != nil && resource.DefaultEncryptionConfiguration != nil && !reflect.DeepEqual(desired.DefaultEncryptionConfiguration, resource.DefaultEncryptionConfiguration) {
+		// Resolve KMS key reference
+		if a.desired.Spec.DefaultEncryptionConfiguration != nil {
+			kmsRef, err := refs.ResolveKMSCryptoKeyRef(ctx, a.reader, a.desired, a.desired.Spec.DefaultEncryptionConfiguration.KmsKeyRef)
+			if err != nil {
+				return err
+			}
+			desired.DefaultEncryptionConfiguration.KmsKeyName = kmsRef.External
+		}
+		resource.DefaultEncryptionConfiguration.KmsKeyName = desired.DefaultEncryptionConfiguration.KmsKeyName
+		updateMask.Paths = append(updateMask.Paths, "default_encryption_configuration")
+	}
+	if !reflect.DeepEqual(desired.IsCaseInsensitive, resource.IsCaseInsensitive) {
+		resource.IsCaseInsensitive = desired.IsCaseInsensitive
+		updateMask.Paths = append(updateMask.Paths, "is_case_sensitive")
+	}
+	if !reflect.DeepEqual(desired.MaxTimeTravelHours, resource.MaxTimeTravelHours) {
+		resource.MaxTimeTravelHours = desired.MaxTimeTravelHours
+		updateMask.Paths = append(updateMask.Paths, "max_time_interval_hours")
+	}
+	if desired.Access != nil && resource.Access != nil && len(desired.Access) > 0 && !reflect.DeepEqual(desired.Access, resource.Access) {
```

Comment:
This means users cannot unset (clear) `Access` list? Not sure if unsetting the entire `Access` list is a valid use case.
Another question is regarding the ordering of this list. Does the underlying API consistently return the list in the same order? If not, we might encounter unexpected diffs even when the user hasn't made any changes to this field.

Reply:
So the behavior is that even if the user leave the access field empty, it will still return with the default access.

Regarding the ordering of the list, there is a sort access method in our mock to give the access list an order, but it is not entirely the same as the real GCP access order. I believe the real GCP returns the object with an access list of a different order consistently. But our test log is consistent with the mock.